mlfinlab/portfolio_optimization/risk_estimators.py

```diff
@@ -0,0 +1,259 @@
+
+        :param corr: (np.array) Correlation matrix
+        :param std: (np.array) Vector of standard deviations
+        :return: (np.array) Covariance matrix
+        """
+
+        cov = corr * np.outer(std, std)
+
+        return cov
+
+    @staticmethod
+    def cov_to_corr(cov):
+        """
+        Derives the correlation matrix from a covariance matrix.
+
+        :param cov: (np.array) Covariance matrix
+        :return: (np.array) Covariance matrix
+        """
+
+        # Calculating standard deviations of the elements
+        std = np.sqrt(np.diag(cov))
+
+        # Transforming to correlation matrix
+        corr = cov / np.outer(std, std)
+
+        # Making sure correlation coefficients are in (-1, 1) range
+        corr[corr < -1], corr[corr > 1] = -1, 1
+
+        return corr
+
+    @staticmethod
+    def _get_pca(hermit_matrix):
+        """
+        Calculates eigenvalues and eigenvectors from a Hermitian matrix. In our case, from the correlation matrix.
+
+        Eigenvalues in the output are on the main diagonal of a matrix.
+
+        :param hermit_matrix: (np.array) Hermitian matrix
+        :return: (np.array, np.array) Eigenvalues matrix, eigenvectors array
+        """
+
+        # Calculating eigenvalues and eigenvectors
+        eigenvalues, eigenvectors = np.linalg.eigh(hermit_matrix)
+
+        # Index to sort eigenvalues in descending order
+        indices = eigenvalues.argsort()[::-1]
+
+        # Sorting
+        eigenvalues = eigenvalues[indices]
+        eigenvectors = eigenvectors[:, indices]
+
+        # Outputting eigenvalues on the main diagonal of a matrix
+        eigenvalues = np.diagflat(eigenvalues)
+
+        return eigenvalues, eigenvectors
+
+    def _denoised_corr(self, eigenvalues, eigenvectors, num_facts):
+        """
+        Shrinks the eigenvalues associated with noise, and returns a de-noised correlation matrix.
+
+        Noise is removed from the correlation matrix by fixing random eigenvalues.
+
+        :param eigenvalues: (np.array) Matrix with eigenvalues on the main diagonal
+        :param eigenvectors: (float) Eigenvectors array
+        :param num_facts: (float) Threshold for eigenvalues to be fixed
+        :return: (np.array) De-noised correlation matrix
+        """
+
+        # Vector of eigenvalues from the main diagonal of a matrix
+        eigenval_vec = np.diag(eigenvalues).copy()
+
+        # Replacing eigenvalues after num_facts to their average value
+        eigenval_vec[num_facts:] = eigenval_vec[num_facts:].sum() / float(eigenval_vec.shape[0] - num_facts)
+
+        # Back to eigenvalues on main diagonal of a matrix
+        eigenvalues = np.diag(eigenval_vec)
+
+        # De-noised covariance matrix
+        cov = np.dot(eigenvectors, eigenvalues).dot(eigenvectors.T)
+
+        # Ne-noised correlation matrix
+        corr = self.cov_to_corr(cov)
+
+        return corr
+
+    def de_noised_cov(self, cov, tn_relation, kde_bwidth):
```

Comment:
Can you change the name of the function to "denoise_covariance()"

Reply:
Done